plugin/trino-spooling-filesystem/src/main/java/io/trino/spooling/filesystem/FileSystemSpoolingManager.java

```diff
@@ -258,18 +205,19 @@ private void checkExpiration(FileSystemSpooledSegmentHandle handle)
         }
     }
 
-    private static String base64Encode(Slice slice)
-    {
-        return Base64.getEncoder().encodeToString(slice.getBytes());
-    }
-
-    private static Slice base64Decode(String base64)
+    private TrinoFileSystem fileSystem(FileSystemSpooledSegmentHandle handle)
     {
-        return wrappedBuffer(Base64.getDecoder().decode(base64));
+        if (handle.encryptionKey().isPresent()) {
+            return new EncryptionEnforcingFileSystem(fileSystem, handle.encryptionKey().get());
+        }
+        return fileSystem;
     }
 
-    private static String sha256Checksum(Slice slice)
+    private static void checkFileExists(TrinoInputFile inputFile)
+            throws IOException
     {
-        return Hashing.sha256().hashBytes(slice.getBytes()).toString();
+        if (!inputFile.exists()) {
+            throw new IOException("Segment not found or expired");
```

Comment:
can be a followup but I think that SpoolingManager should throw TrinoException. Not IOException. Then you can provide nice error code which explains what is going on.

Reply:
Yeah, that area can be improved